repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletResponseWrapper.scala

```diff
@@ -46,84 +53,253 @@ import javax.servlet.http.HttpServletResponse
   extends javax.servlet.http.HttpServletResponseWrapper(originalResponse)
   with HeaderInteractor {
 
+  object ResponseBodyType extends Enumeration {
+    val Available, OutputStream, PrintWriter = Value
+  }
+
+  private val bodyOutputStream = bodyMode match {
+    case ResponseMode.PASSTHROUGH => new PassthroughServletOutputStream(desiredOutputStream)
+    case ResponseMode.READONLY => new ReadOnlyServletOutputStream(desiredOutputStream)
+    case ResponseMode.MUTABLE => new MutableServletOutputStream(desiredOutputStream)
+  }
+
+  private val bodyPrintWriter = new PrintWriter(bodyOutputStream)
+  private val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+
+  private var headerMap: Map[String, Seq[String]] = new TreeMap[String, Seq[String]]()(caseInsensitiveOrdering)
+  private var responseBodyType = ResponseBodyType.Available
+
   /**
-   * This constructor chains to the main constructor using the original responses output stream  as the last arguement.
+   * This constructor chains to the main constructor using the original responses output stream  as the last argument.
    *
    * @constructor the constructor to use when you don't plan on using a custom stream for processing
    * @param originalResponse the response to be wrapped
-   * @param headerMode the mode to use for header accessiblity
+   * @param headerMode the mode to use for header accessibility
    * @param bodyMode the mode to use for body accessibility
    */
   def this(originalResponse: HttpServletResponse, headerMode: ResponseMode, bodyMode: ResponseMode) =
     this(originalResponse, headerMode, bodyMode, originalResponse.getOutputStream)
 
-  override def getHeaderNamesList: util.List[String] = ???
+  override def getHeaderNamesList: util.List[String] = getHeaderNames.toList
+
+  override def getHeaderNames: util.Collection[String] = headerMap.keys
+
+  override def containsHeader(name: String): Boolean = headerMap.contains(name)
+
+  override def getHeader(name: String): String = getHeaderValues(name).headOption.orNull
+
+  override def getHeaders(name: String): util.Collection[String] = getHeaderValues(name)
+
+  override def getHeadersList(name: String): util.List[String] = getHeaderValues(name)
+
+  override def getPreferredHeaders(name: String): util.List[String] = getPreferredHeaderValues(getHeaderValues(name))
+
+  override def getPreferredSplittableHeaders(name: String): util.List[String] =
+    getPreferredHeaderValues(getSplittableHeaders(name))
+
+  override def getSplittableHeaders(name: String): util.List[String] =
+    getHeaderValues(name).foldLeft(List.empty[String])((list, value) => list ++ value.split(","))
+
+  /**
+   * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
+   */
+  override def addHeader(name: String, value: String): Unit = {
```

Comment:
This behavior isn't correct adds should always work, it's just a question of whether they immediately push through.